README.md
<!--
This README describes the package. If you publish this package to pub.dev,
this README's contents appear on the landing page for your package.

For information about how to write a good package README, see the guide for
[writing package pages](https://dart.dev/guides/libraries/writing-package-pages).

For general information about developing packages, see the Dart guide for
[creating packages](https://dart.dev/guides/libraries/create-library-packages)
and the Flutter guide for
[developing packages and plugins](https://flutter.dev/developing-packages).
-->
A package to add google pay and apple pay to your Flutter application.
It is the payment package for the stage 5 task of HNGx internship.

## Platform Support
| Android  | Google Pay |
|:---------:|:----------:|
|   iOS:    | Apple Pay  |

## Features

This package helps to integrate Google pay and Apple pay into your app.

## Getting started


## Usage

1. Import the package into your project.
   To start using this plugin, add `in_app_payment` as a [dependency in your pubspec.yaml file](https://flutter.io/platform-plugins/):
   ```yaml
   dependencies:
   in_app_payment: ^0.0.2
   ```

2. Initialize HNGPay:
   ```dart
      final pay = HNGPay();
   ```

3. For Android, invoke the googlePay function in place of your 'pay' button in your app and pass the amount to be paid as argument.
   ```dart
   pay.googlePay(context, amountToPay: "24", userID: '23')
   ```
4. For IOS, invoke the applePay function in place of your 'pay' button in your app and pass the amount to be paid as argument.
    ```dart
    pay.applePay(context, amountToPay: "24", userID: '23'),
    ```

### Example
This snippet assumes the existence a payment configuration
   ```dart
   import 'package:flutter/material.dart';
   import 'package:in_app_payment/in_app_payment.dart';
   
   void main() {
     runApp(const MyApp());
   }
   
   class MyApp extends StatelessWidget {
     const MyApp({super.key});
     @override
     Widget build(BuildContext context) {
       return MaterialApp(
         title: 'Flutter Demo',
         theme: ThemeData(
           colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
           useMaterial3: true,
         ),
         home: const MyHomePage(title: 'Flutter Demo Home Page'),
       );
     }
   }
   
   class MyHomePage extends StatefulWidget {
     const MyHomePage({super.key, required this.title});
   
     final String title;
   
     @override
     State<MyHomePage> createState() => _MyHomePageState();
   }
   
   class _MyHomePageState extends State<MyHomePage> {
       
       final pay = HNGPay();
   
     @override
     Widget build(BuildContext context) {
       return Scaffold(
         appBar: AppBar(
           backgroundColor: Theme.of(context).colorScheme.inversePrimary,
           title: Text("Payment test"),
         ),
         body: Center(
           child: Container(
             child: pay.googlePay(context, amountToPay: "24", userID: '23'), //for android implementaion
           ),
         ),
       );
     }
   }
   
   ```

## Authors
- [Rayhan Awojobi](https://github.com/dexterousnorthpaw)
- [Ridwan Lawal](https://github.com/Reazy-ai)
- [Zaccheus Oluwole](https://github.com/Zacchaeus-Oluwole)


CHANGELOG.md
## 0.0.1

Google pay button and Apple pay button are created so that when click it opens up the payment pop up
of any of the buttons picked.

## 0.0.2

Update to Google pay and Apple Pay buttons now required user ID to be able to update payment for the
current user.

test/in_app_payment_test.dart
import 'package:flutter_test/flutter_test.dart';

<<<<<<< HEAD
void main() {}
=======
void main() {}
>>>>>>> ef2384c66f66e13310369bc38a581e67fb81d91c


lib/buttons/pay_buttons.dart
import 'package:flutter/material.dart';
import 'package:in_app_payment/configurations/payment_config.dart';
import 'package:pay/pay.dart';

class HNGPay {
  /// Displays a google pay button on Android device to handle payment
  /// And Sends the userID and token to the backend for payment processing
  googlePay(BuildContext context,
      {required String amountToPay, required String userID}) {
    return GooglePayButton(
      loadingIndicator: const CircularProgressIndicator(),
      paymentConfiguration:
          PaymentConfiguration.fromJsonString(AppPayment.googlePayConfig),
      onPaymentResult: (result) {
        //TODO : also send to backend
        debugPrint(result.toString());
        ScaffoldMessenger.of(context).showSnackBar(
            const SnackBar(content: Text("Processing Payment, Thank you")));

        ///SnackBar to show payment processing
      },
      onError: (err) {
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(
            content: Text(
              err.toString(),
            ),
          ),
        );
      },
      paymentItems: [
        PaymentItem(
          amount: amountToPay,
          status: PaymentItemStatus.final_price,
        ),
      ],
    );
  }

  /// Displays an apple pay button on IOS device to handle payment
  /// And Sends the userID and token to the backend for payment processing
  applePay(BuildContext context,
      {required String amountToPay, required String userID}) {
    return ApplePayButton(
      loadingIndicator: const CircularProgressIndicator(),
      paymentConfiguration: PaymentConfiguration.fromJsonString(
        AppPayment.applePayConfig,
      ),
      paymentItems: [
        PaymentItem(
          amount: amountToPay,
          status: PaymentItemStatus.final_price,
        ),
      ],
      onPaymentResult: (result) {
        // TODO: Send the result to the backend.
        debugPrint(result.toString());
        ScaffoldMessenger.of(context).showSnackBar(
          const SnackBar(
            content: Text("Processing Payment, Thank you"),
          ),
        );
      },
      onError: (err) {
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(
            content: Text(
              err.toString(),
            ),
          ),
        );
      },
    );
  }
}


lib/buttons/google_pay_button.dart
// ignore_for_file: camel_case_types

// import 'dart:convert';

import 'package:flutter/widgets.dart';
import 'package:in_app_payment/configurations/payment_config.dart';
import 'package:pay/pay.dart';
// import 'package:http/http.dart' as http;



import 'package:flutter/material.dart';

googlePay({required String amountToPay}) {
  return Container(
    child: GooglePayButton(
        loadingIndicator: CircularProgressIndicator(),
        paymentConfiguration: PaymentConfiguration.fromJsonString(AppPayment.googlePayConfig),
        onPaymentResult: (result) {
          print(result);
        },
        paymentItems: [
          PaymentItem(amount: amountToPay, status: PaymentItemStatus.final_price)
        ]),
  );
}


// Future<PaymentResult> sendGooglePaymentResult( String paymentResult) async {
//   final response = await http.post(
//     Uri.parse('https://api/'),
//
//     headers: <String, String>{
//       'Content-Type': 'application/json; charset=UTF-8',
//     },
//     body: jsonEncode(<String, String>{
//       // 'title': title,
//       'paymentResult': paymentResult,
//
//     }),
//   );
//
//   if (response.statusCode == 201) {
//     // If the server did return a 201 CREATED response,
//     // then parse the JSON.
//     // print(response.statusCode);
//     // print(response.body);
//     return PaymentResult.fromJson(jsonDecode(response.body));
//   } else {
//     // If the server did not return a 201 CREATED response,
//     // then throw an exception.
//     throw Exception('Failed to create payment result.');
//   }
// }
//
//
//
// class PaymentResult {
//   final String paymentResult;
//
//   PaymentResult({
//     required this.paymentResult,
//   });
//
//   factory PaymentResult.fromJson(Map<String, dynamic> json) {
//     return PaymentResult(
//       paymentResult: json['paymentResult'],
//     );
//   }
// }
//
//
//
// class googlePayButton extends StatefulWidget {
//   final String itemName;
//   final String amountToPay;
//
//   const googlePayButton({super.key, required this.itemName, required this.amountToPay});
//
//   @override
//   // ignore: no_logic_in_create_state
//   State<googlePayButton> createState() => _googlePayButtonState(itemName, amountToPay);
// }
//
// class _googlePayButtonState extends State<googlePayButton> {
//   final String itemName;
//   final String amountToPay;
//
//   _googlePayButtonState(this.itemName, this.amountToPay);
//
//
//   // late String newitemName;
//   // late String newamountToPay;
//
//   // @override
//   // void initState(){
//   //   super.initState();
//   //   newitemName = widget.itemName;
//   //   newamountToPay = widget.amountToPay;
//   // }
//   @override
//   Widget build(BuildContext context) {
//     return GooglePayButton(
//       paymentConfiguration:
//           PaymentConfiguration.fromJsonString(AppPayment.googlePayConfig),
//       onPaymentResult: onGooglePaymentResult,
//       paymentItems: [
//         PaymentItem(
//           label: itemName,
//           amount: amountToPay,
//           status: PaymentItemStatus.final_price,
//         ),
//       ]
//     );
//   }
// }
//
//
// void onGooglePaymentResult(paymentResult) {
//     debugPrint(paymentResult.toString());
//     // TODO: Send the result to the server.
//   }

lib/configurations/payment_config.dart
//dummy config data from the docs pending when we get access to a google business
// account and an apple pay merchant account

class AppPayment {
  //class to hold the configuration details
  AppPayment._();

  ///configuration details for google pay
  //TODO: change merchant details
  static String googlePayConfig = '''{
  "provider": "google_pay",
  "data": {
    "environment": "TEST",
    "apiVersion": 2,
    "apiVersionMinor": 0,
    "allowedPaymentMethods": [
      {
        "type": "CARD",
        "tokenizationSpecification": {
          "type": "PAYMENT_GATEWAY",
          "parameters": {
            "gateway": "example",
            "gatewayMerchantId": "gatewayMerchantId"
          }
        },
        "parameters": {
          "allowedCardNetworks": ["VISA", "MASTERCARD"],
          "allowedAuthMethods": ["PAN_ONLY", "CRYPTOGRAM_3DS"],
          "billingAddressRequired": true,
          "billingAddressParameters": {
            "format": "FULL",
            "phoneNumberRequired": true
          }
        }
      }
    ],
    "merchantInfo": {
      "merchantId": "01234567890123456789",
      "merchantName": "Example Merchant Name"
    },
    "transactionInfo": {
      "countryCode": "US",
      "currencyCode": "USD"
    }
  }
}''';

  //configuration for apple pay
// TODO: Change merchant details
  static String applePayConfig = '''{
  "provider": "apple_pay",
  "data": {
    "merchantIdentifier": "merchant.com.sams.fish",
    "displayName": "Sam's Fish",
    "merchantCapabilities": ["3DS", "debit", "credit"],
    "supportedNetworks": ["amex", "visa", "discover", "masterCard"],
    "countryCode": "US",
    "currencyCode": "USD",
    "requiredBillingContactFields": ["emailAddress", "name", "phoneNumber", "postalAddress"],
    "requiredShippingContactFields": [],
    "shippingMethods": [
      {
        "amount": "0.00",
        "detail": "Available within an hour",
        "identifier": "in_store_pickup",
        "label": "In-Store Pickup"
      },
      {
        "amount": "4.99",
        "detail": "5-8 Business Days",
        "identifier": "flat_rate_shipping_id_2",
        "label": "UPS Ground"
      },
      {
        "amount": "29.99",
        "detail": "1-3 Business Days",
        "identifier": "flat_rate_shipping_id_1",
        "label": "FedEx Priority Mail"
      }
    ]
  }
}''';
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:in_app_payment/in_app_payment.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  ///initializing payment package
  final pay = HNGPay();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Theme.of(context).colorScheme.inversePrimary,
        title: Text("Payment test"),
      ),
      body: Center(
        child: SizedBox(
          child: pay.googlePay(context, amountToPay: "24", userID: '23'),

          /// accepting the amount and user id
        ),
      ),
    );
  }
}
